pkg/cmd/roachtest/registry/operation_spec.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/operation"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+)
+
+// OperationDependency specifies what an operation requires from a cluster to
+// be able to run. This type is hierarchical; OperationRequiresPopulatedDatabase
+// implies OperationRequiresDatabaseSchema, OperationRequiresSqlCOnnection etc.,
+// i.e. the cluster must running  cockroach processes with dbs/tables created
+// and populated with some data.
+type OperationDependency int
+
+const (
+	OperationRequiresNodes OperationDependency = iota
+	OperationRequiresSqlConnection
+	OperationRequiresDatabaseSchema
+	OperationRequiresPopulatedDatabase
+)
+
+// OperationSpec is a spec for a roachtest operation.
+type OperationSpec struct {
+	Skip string // if non-empty, operation will be skipped
+
+	Name string
+	// Owner is the name of the team responsible for signing off on failures of
+	// this operation that happen in the release process. This must be one of a limited
+	// set of values (the keys in the roachtestTeams map).
+	Owner Owner
+	// The maximum duration the operation is allowed to run before it is considered
+	// failed. This timeout applies _individually_ to Run() and Cleanup() each, but
+	// does not include any scheduling waits for either method.
+	Timeout time.Duration
+
+	// CompatibleClouds is the set of clouds this test can run on (e.g. AllClouds,
+	// OnlyGCE, etc). Must be set.
+	CompatibleClouds CloudSet
+
+	// Dependency specify the types of resources required for this roachtest
+	// operation to work. This will be used in filtering eligible operations to
+	// run. Note that this field is strictly heirarchical; requiring a populated
+	// db assumes that this
+	//
+	// TODO(bilal): Unused.
+	Dependency OperationDependency
+
+	// CanRunConcurrently specifies whether this operation is safe to run
+	// concurrently with other operations that have CanRunConcurrently = true. For
+	// instance, a random-index addition is safe to run concurrently with most
+	// other operations like node kills, while a drop would need to run on its own
+	// and will have CanRunConcurrently = false.
+	//
+	// TODO(bilal): Unused.
+	CanRunConcurrently bool
+
+	// Run is the operation function.
+	Run func(ctx context.Context, o operation.Operation, c cluster.Cluster)
+
+	// CleanupWaitTime is the min time to wait before running the Cleanup method.
+	CleanupWaitTime time.Duration
+
+	// Cleanup is the operation cleanup function, if defined.
+	Cleanup func(ctx context.Context, o operation.Operation, c cluster.Cluster)
+}
+
+// TestSpec() converts this operation to a TestSpec for use with roachtest run-operation.
+// This adapter is only meant for operation debugging purposes and is not the
+// canonical approach for referring to an Operation spec, as it drops many
+// operation-specific fields.
+func (s *OperationSpec) TestSpec() TestSpec {
```

Comment:
Are we still planning on using the existing test runner? or would it make sense to just have a new simple runner for operations? My concern is that this complicates the existing runner by having to conditionally do things differently for operations and tests. One thing that comes to mind is how this spec will prevent cluster deletion, or collecting logs etc.

Reply:
Yeah, logs are the big question mark for me if we stop using the existing runner. But seeing as there's a preference for not reusing the existing runner, I've stripped this out for now and will roll a new `operation.Operation` implementation in the next PR.